### PR TITLE
WIP: editorial pass to rearrange/layer spec

### LINKED ISSUE #283: Layering proposal for v3
## Context

Currently, the PE spec is complex for at least 3 reasons:
1. It's hard to detangle which objects are needed for simple vs advanced use cases. An example of an advanced use case is submission requirements
2. Elements such as `subject_is_holder`, `subject_is_issuer`, `same_subject`, and `statuses` risk fragmentation. They are left up to implementors, and currently lack sufficient examples needed to determine fitness and adequacy for interoperability. We also don't have example implementations for these.
3. In attempting to be format-agnostic, it loses opportunities for clarity (potentially risking interoperability). 

2 and 3 overlap -- in the context of a particular format, it's possible to be more precise about how the subject, holder, and issuer entities are determined.

These problems (and more) are what's leading to (what is now) a trend of creation of PE profiles, in which implementers define a PE subset that is implementable/testable. However, some of these profiles (targeted at the simple/80% use cases) look very much the same. David Chadwick's [PE profile for OIDC](https://hackmd.io/5b60vH2mSXevoQeAbmd9Lw) is very close to the subset I have implemented, and (from what I understand) others have too. 

This presents an opportunity to simplify the spec and improve interoperability. (Continuing to define profiles that intersect heavily, but differ slightly -- without work to unify the common intersection -- will result in fragmentation.) In v3 we should invert the approach, applying these simplifications to the spec itself and clarifying the role profiles will take

## Proposed Presentation Exchange v3 Layers / Features

- Base layer = scaled-down version of Input Descriptors (see next section)
    - In this layer, the only constraints are field constraints (in fact, note that it's collapsed in the diagram)
- Submission requirements
- Collection of layers/features to address everything not included above -- additional constraints, predicates, etc
    - everything in this category needs to be revised for the reasons described in "context"
    - subject_is_issuer, etc may make sense to add in the context of format-specific considerations 

TBD: 
- terminology ("layer" vs "feature")
- spec structure (separate docs vs 1)

## Draft (WIP)

See [hackmd draft (WIP)](https://hackmd.io/wniK3529RieDxkGOTp5J8g)

### Diagram of resulting base layer

image

Note: only a few differences from David's profile:
- added `format` to input_descriptor per #279 
- moved limit_disclosure out of this layer

**Claim**: this corresponds to an implementable base layer that we can immediately add a reference implementation and test vectors for. Exciting!

#### Break up `constraints` object

The current constraints object is especially confusing. It contains a mix of constraint types, some of which apply to an entire input, some of which apply to fields, and one which appears to apply across inputs (therefore possibly not belonging here -- see `same_subject`). 

We can consider breaking this up (as I've done above with field_contraints) to make this more precise.

#### Add back remaining properties prioritized by ability to define in a testable manner  

We can proceed by working through properties where we have specific examples and interoperability requirements. `subject_is_issuer` is a good example:
- OIDC has a special reserved issuer value for this, and an implementor would check something like `issuer = self-issued.me`
- A VC-driven approach might prefer to check that the issuer (id) field equals the credential subject (id) field

We can define a way that allows PE or PE profiles to define how these are checked.

### PR DESCRIPTION
**Pull request by kimdhamilton**
[See preview](https://centrehq.github.io/presentation-exchange/build/)

Reflects discussion of Jan 20 Discussion. 

**Notes**
1. `limit_disclosure` is now entirely moved to a Feature
2. Rolled logically-similar properties more aggressively into single Features (e.g., Advanced Constraints)
3. Added language to address MUST/MAY topic, e.g., "When using this Feature, the constraints object MAY contain a limit_disclosure property. If present, its value MUST be one of the following strings"
4. I think there still may be some lingering weirdness in Submission Requirements, i.e., if a Feature introduces extensions to multiple objects
5. Tried to add special formatting to Input Evaluation Extension handling. It's awkward and I probably need to discuss with @csuwildcat 

**Issues**
1. Other similar restructuring improvements: consider presentation submission special case (for 1) to avoid need for `descriptor_map`?  Consider doing this in separate PR
2. Need language around processing entity handling when Features not supported, reflecting the "naughty behavior" discussion. i feel like I can't capture this as well as @csuwildcat 
3. TODO: I need to ensure David's examples are captured
4. TODO I manually checked in index.html in my branch; will remove that when merging
5. TODO: Update authors/editors. Add David Chadwick and get updated list in general.

Feel free to refer to Note # or Issue # when commenting. E.g., "Note 2: I disagree"

Fixes #283; Overrides #275

### CODE REVIEW

**Comment by bumblefudge** [comment]
NB @David-Chadwick

**Comment by bumblefudge** [comment]
I've been trying to understand Issue # 2, and i am a little stumped.  Part of me thinks we should spill a paragraph's worth of free digital ink after the second graf of `## Input Evaluation` (L 1191 in the PR), explaining that the choices taken by processors in Input Evaluation may be judged or tracked by Verifiers, and to consider optional features subject to reputation-scoring.  Was this what Daniel meant?  Would a mentioned in the definition of Features in the `## Terminology` section help, since they are free from conformance-stick but still subject to the reputation-carrot?

**Comment by bumblefudge** [comment]
Sidenote, are these 5 issues being opened as tracking issues if not addressed by this merge? Not sure if that was implied or not but asking for the sake of clear process.

**Comment by kimdhamilton** [comment]
> Sidenote, are these 5 issues being opened as tracking issues if not addressed by this merge? Not sure if that was implied or not but asking for the sake of clear process.

It is a mix of "notes to self" that I will address before merging (e.g. index.html hack), issues we may quickly decide on as a group (and I can address with this change", and (default) issues I will create as separate tracking issues if not address.

**Comment by kimdhamilton** [comment]
> I've been trying to understand Issue # 2, and i am a little stumped. Part of me thinks we should spill a paragraph's worth of free digital ink after the second graf of `## Input Evaluation` (L 1191 in the PR), explaining that the choices taken by processors in Input Evaluation may be judged or tracked by Verifiers, and to consider optional features subject to reputation-scoring. Was this what Daniel meant? Would a mentioned in the definition of Features in the `## Terminology` section help, since they are free from conformance-stick but still subject to the reputation-carrot?

Definitely that's the right line of thinking -- this is just an area I hadn't already addressed in the PR but something I want tracked. So worst case, if I don't have time to get to it in this already massive change, we can create a tracking issue

File: spec/spec.md
Comment (JaceHensley) [inline spec/spec.md:932]: #supernit maybe the order of `"preferred"` and `"required"` should be the same for properties that follow that pattern?

**Review by JaceHensley — COMMENTED** [review]
Issue #1: Can you explain this one more for me? Are you saying there are cases where a `descriptor_map` isn't needed? The only case I can think of that could omit that would be when requesting a single credential but even then I think you'd need to know where to look for the credential in the envelop

File: spec/spec.md
Comment (bumblefudge) [reply spec/spec.md:932]: that's not a supernit that's a valid question, i'd say

**Comment by JaceHensley** [comment]
Note # 1: Based on discussion on Jan 27th `limit_disclosure` should be moved into the base profile with additional language around `"required"` ("If you don't support "required" don't return anything" / "Wallets must understand how to at least know when "limit_disclosure" is "required"")

File: spec/spec.md
Comment (kimdhamilton) [reply spec/spec.md:926]: Yeah, i'm putting this on hold for now

**Comment by kimdhamilton** [comment]
Thanks for the feedback! Here are the updates:

- Addressed feedback from Jan 27
- Renamed the now-predicate section to "Predicate Feature"

[See preview](https://centrehq.github.io/presentation-exchange/build/)

I agree with the discussion point that we should land this and follow up with subsequent PRs.

Things that are a prereq for check-in (for my tracking):
- ensure index.html is not in merge

Issues I will create:
- Need a better description of feature behavior, i.e., "light up" if features understood, backwards compatible
- Need language around processing entity handling when Features not supported, reflecting the "naughty behavior" discussion. @csuwildcat 
- (Unless we're ok with it as is) Fix formatting in "Input Evaluation" section for Features.
- Update authors/editors

Other group TODOs:
- review David's original PR and determine if there's anything missing from my check-in and create issues

File: spec/spec.md
Comment (nklomp) [inline spec/spec.md:1034]: What is the use case for a preferred subject_is_issuer from a Verifier perspective? If I don't care I would not put the constraint in to begin with as it is a MAY.

I do see a use case for the inverse of required, basically stating that a subject is not the issuer.

I guess this comment is applicable to all relational constraints.

File: spec/spec.md
Comment (nklomp) [inline spec/spec.md:1163]: Does it need mentioning that this depends on `limit_disclosure` being preferred or required?

**Review by nklomp — COMMENTED** [review]
Some small remarks/questions.

Unfortunately I couldn't attend last meeting, so hopefully not a stupid remark/question.
What is the exact use case for preferred relational constraints that are optional to begin with?

For selective disclosure the use case is really obvious from a Verifier Perspective. _I want to receive as little information as possible if you support selective disclosure, but I am okay with getting everything if you don't support it._

How would that work for subject_is_holder as example?. Basically you would be stating, "I don't care", which you could also express by not including it in the first place as it is optional.

File: spec/spec.md
Comment (bumblefudge) [reply spec/spec.md:1163]: It might help the LD-unfamiliar reader to add a sentence like: 
`LD framing objects are required for some forms of selective disclosure presentation.`
I'm not sure it is precise to describe it as a dependence in either direction, though-- you might want to include a framing object regardless of selective disclosure, just because you like RDF. who knows why people do the things they do!

File: spec/spec.md
Comment (kimdhamilton) [reply spec/spec.md:199]: Thanks @nklomp, these are good questions, and we may need to bring these in as follow-up issues (we're trying to control the size/scope of this PR). Let's discuss tomorrow and figure out if it can be easily addressed in this PR or if it needs to be tracked in a separate issue

File: spec/spec.md
Comment (nklomp) [reply spec/spec.md:1163]: Yeah agree. Although you mention `some forms of selective disclosure` already, maybe best to replace `required' with `are being`, to make it more clear that frames typically are optional even in case of selective disclosure in PEX

File: spec/spec.md
Comment (nklomp) [reply spec/spec.md:1034]: Yeah, but then you are expressing it at a per VC level anyway right? I guess it would make it a bit more expressive, but if I create a definition that simply leaves out the `preferred` for the relational constraint, I would have exactly the same constraint as it is only expressing purpose/intent.

File: spec/spec.md
Comment (brentzundel) [inline spec/spec.md:615]: would it work to nest all of the features into their own section?

File: spec/spec.md
Comment (brentzundel) [reply spec/spec.md:1034]: iirc, the original set of possible values was 'required' and 'allowed', but it was changed to 'preferred' to match the other properties.
'allowed' makes much more sense as a possible option, i.e., I need to know your address and it can come some issuer or from yourself.

**Review by brentzundel — COMMENTED** [review]
overall this looks really great. I like the new structure and appreciate the work put in to make the spec better.

File: spec/spec.md
Comment (kimdhamilton) [reply spec/spec.md:1034]: @brentzundel please create an issue

File: spec/spec.md
Comment (kimdhamilton) [reply spec/spec.md:200]: @nklomp please create issue

File: spec/spec.md
Comment (kimdhamilton) [reply spec/spec.md:616]: TODO @kimdhamilton

File: spec/spec.md
Comment (kimdhamilton) [reply spec/spec.md:580]: TODO @kimdhamilton: move to right before json schemas

File: spec/spec.md
Comment (kimdhamilton) [reply spec/spec.md:616]: TODO @kimdhamilton

**Review by brentzundel — APPROVED** [review]

**Pull request merged**
